### PR TITLE
Android java bindings

### PR DESCRIPTION
**Pull request by clemens-msupply**
- Adds a package for Android java bindings.
- Refactors a start_server method that can be used in main and from Android
- Add "off_switch" parameter to start_server to shutdown the server (e.g. when running on Android)
- Remove unwrap from sync (so that the error can be logged on Android)

### CODE REVIEW

File: server/src/lib.rs [hunk 1]
```diff
@@ -1,7 +1,127 @@
...
 pub mod environment;
 pub mod middleware;
 pub mod settings;
 pub mod sync;
 pub mod test_utils;
+
+/// Starts the server
+///
+/// This method doesn't return until a message is send to the off_switch.
+pub async fn start_server(
```
Comment (andreievg) [inline server/src/lib.rs:37]: is this copy paster straight from main.rs ? or any other changes here ?

File: server/src/sync/connection.rs [hunk 2]
```diff
@@ -71,20 +71,20 @@ pub struct SyncConnection {
 }
 
 impl SyncConnection {
-    pub fn new(settings: &SyncSettings) -> SyncConnection {
+    pub fn new(settings: &SyncSettings) -> anyhow::Result<SyncConnection> {
```
Comment (andreievg) [inline server/src/sync/connection.rs:74]: would this error only happen if Url parse fail ?

File: android/src/android.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                    },
+                    database: DatabaseSettings {
+                        username: "n/a".to_string(),
+                        password: "n/a".to_string(),
+                        port: 0,
+                        host: "n/a".to_string(),
+                        database_name: db_path,
+                    },
+                    sync: SyncSettings {
+                        url: "http://localhost".to_string(),
```
Comment (andreievg) [inline android/src/android.rs:91]: should this be a parameter ? Btw i think this sort of settings should be stored in DB, makes it easier to handle between platforms then toml file

File: android/src/android.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                        username: "username".to_string(),
+                        password: "password".to_string(),
+                        interval: 300,
+                    },
+                    auth: AuthSettings {
+                        token_secret: "Make me configurable".to_string(),
```
Comment (andreievg) [inline android/src/android.rs:97]: todo

**Review by andreievg — APPROVED** [review]
Looks good.

But since the makefile, android.rs and host.rs are quite technical, can we have in depth documentation about this please.? 

Liike why is there darwin host option for android build, what is server_bucket.

File: server/src/lib.rs (hunk 1 above)
Comment (clemens-msupply) [reply server/src/lib.rs:37]: Mostly there is only an off_switch channel added to stop the server programmatically

File: server/src/sync/connection.rs (hunk 2 above)
Comment (clemens-msupply) [reply server/src/sync/connection.rs:74]: yes I was debugging and the panic don't show up in the Android logs so I needed the error to write it to the log...

File: android/src/android.rs (hunk 4 above)
Comment (clemens-msupply) [reply android/src/android.rs:105]: Done [d806b14](https://github.com/openmsupply/remote-server/pull/698/commits/d806b14d8affd1bd2af6b01edc8acb728c68608c)

**Comment by clemens-msupply** [comment]
I added more comments to host.sh, makefile, and android.rs

**Pull request merged**